Adds default namespace to schema.

garex opened the pull request:
Adds default namespace to schema

Some databases supports namespaces and has default one. For example Postgres has default "public".

When schema created from classes, which has table names without namespaces, but this namespace is assumed, we should add it explicitly, if it not exists yet.

It will help in diffs between real DB and config shemas.

The visitor is skipping platforms that don't support schemas.

This is 1st part of fix DBAL-1168 #1110

garex committed:
Adds default namespace to schema.

Some databases supports namespaces and has default one. For example
Postgres has default "public".

When schema created from classes, which has table names without
namespaces, but this namespace is assumed, we should add it explicitly,
if it not exists yet.

It will help in diffs between real DB and config shemas.

This is 1st part of fix DBAL-1168

garex commented:
Next step will be to add this visitor to the end of `Doctrine\ORM\Tools\SchemaTool::getSchemaFromMetadata`

    if ($this->platform->supportsSchemas()) {
        $schema->visit(new AddDefaultNamespace($this->platform));
    }

Ocramius commented on lib/Doctrine/DBAL/Schema/Visitor/AddDefaultNamespace.php:
```diff
@@ -0,0 +1,58 @@
+ */
+class AddDefaultNamespace extends AbstractVisitor
+{
+
```
Docblock

garex replied:
What wrong with it?

Ocramius replied:
It's missing :-)

garex replied:
> It's missing :-)

What is missing?? There is top comment and class comment. What is "docblock" then? Can you provide an example?

teohhanhui replied:
``` php
/**
 * @var AbstractPlatform
 */
private $platform;
```

Ocramius commented on tests/Doctrine/Tests/DBAL/Schema/Visitor/AddDefaultNamespaceTest.php:
```diff
@@ -0,0 +1,82 @@
+    {
+        return [
+            [new PostgreSqlPlatform()],
+            [new SQLServerPlatform()],
```
What about platforms with no schema support?

garex replied:
It will not work there as it should be added only if platform support default namespace. You think it should be added in ORM without such check and do this check inside itself?

Ocramius replied:
As discussed later, the check should probably be done by the visitor, skipping platforms that don't support schemas.

Ocramius commented:
> Next step will be to add this visitor to the end of `Doctrine\ORM\Tools\SchemaTool::getSchemaFromMetadata`

Could be added by default, but then act as no-op if the platform doesn't support schemas?

garex commented:
> Could be added by default, but then act as no-op if the platform doesn't support schemas?

I also thought about this option first, but then deside to do the check outside to give more readability at the place where it should be added at `getSchemaFromMetadata`.

But if your experience tells you to move it inside -- let's move it inside. It will become more reusable then.

Ocramius commented:
If it's not a no-op when the platform does not support schemas, then it will become a setup issue for most users, making working apps not portable. IMO, it should be done inside the visitor.

garex commented:
Ok, I will rework it. Pls not merge many commits, I will rebase em after approve. 

Regarding docblock. What bad there? Bad `@since` or description?

Ocramius commented on lib/Doctrine/DBAL/Schema/Visitor/AddDefaultNamespace.php:
```diff
@@ -0,0 +1,58 @@
+ * It will help in diffs between real DB and config shemas.
+ *
+ * @author Alexander Ustimenko <a@ustimen.co>
+ * @since  2.5
```
2.6

garex committed:
!!! TRASH COMMIT NOT FOR MERGE !!!
--- Branch will be rebased after approve ---
Merge branch 'master' into DBAL-1168-add-default-schema-name

garex committed:
.. Add docblocks, fix since, wrap text to 80 chars

garex committed:
.. make visitor noop on incompatible platforms

garex committed:
.. Test incompatible platforms unaffected

garex committed:
.. Fix imports in test

garex commented:
Pls give me know when you think it will be ready for merge -- I will rebase it. Or you can rebase it on master and squash all in one commit.

garex commented:
@deeky666 when approximately will you plan to review?

garex commented:
@Ocramius please assing someone else or review by yourself.

Two weeks passed already and @deeky666 is not reponds.

phansys commented on lib/Doctrine/DBAL/Schema/Visitor/AddDefaultNamespace.php:
```diff
@@ -0,0 +1,71 @@
+{
+
+    /**
+     * @var \Doctrine\DBAL\Platforms\AbstractPlatform
```
This class is already imported, so you should use `@var AbstractPlatform` instead.

garex replied:
Not sure. In other classes in same directory also fully qualified names used. So I decided to leave it as it was here before.

phansys commented on lib/Doctrine/DBAL/Schema/Visitor/AddDefaultNamespace.php:
```diff
@@ -0,0 +1,71 @@
+ */
+class AddDefaultNamespace extends AbstractVisitor
+{
+
```
Please, remove surplus LF.

garex committed:
!!! TRASH COMMIT NOT FOR MERGE !!!
--- Branch will be rebased after approve ---
Merge branch 'master' into DBAL-1168-add-default-schema-name

garex committed:
.. Remove surplus LF.

garex commented:
This PR is covered with cobwebs. Betting, it will be merged may be after New Year.

![Cobwebs](http://s1.1zoom.ru/big3/493/PuppeteerSpider_web_Doll_462236.jpg)

Ocramius commented:
There's waaaaaay older stuff around here. If @deeky666 didn't have time to check it, he's evidently busy at the moment.

garex commented:
@Ocramius As I see, @deeky666 has not any activity on his github profile after 2016, July.

Do you know him? Is he alive/healthy?

Can you assign someone else from 31 people that are in Doctrine's members team?

Ocramius commented:
No, I cannot. He's pretty much the gatekeeper, and I talk with him from
time to time, he just is that busy IRL ATM.

On 8 Nov 2016 8:25 a.m., "Alexander Ustimenko" <notifications@github.com>
wrote:

@Ocramius <https://github.com/Ocramius> As I see, @deeky666
<https://github.com/deeky666> has not any activity on his github profile
after 2016, July.

Do you know him? Is he alive/healthy?

Can you assign someone else from 31 people that are in Doctrine's members
team?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/doctrine/dbal/pull/2490#issuecomment-259064881>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAJakPH_vngo2oVNCiUqDOpfF0l53kzRks5q8CPrgaJpZM4JuIOV>
.

phansys commented on lib/Doctrine/DBAL/Schema/Visitor/AddDefaultNamespace.php:
```diff
@@ -0,0 +1,70 @@
+     */
+    public function acceptSchema(Schema $schema)
+    {
+        if (!$this->platform->supportsSchemas()) {
```
Please, respect Doctrine CS.

teohhanhui replied:
Ii'm guessing it's this one? https://github.com/doctrine/coding-standard/pull/7

> Logical NOT operators (!) MUST have leading and trailing spaces

garex replied:
Holy shit, pettifoggers )) I will README https://github.com/doctrine/coding-standard and will come back to you soon.

phansys reviewed (changes requested):
Status: Needs work.

garex committed:
!!! TRASH COMMIT !!! not for future merge

Merge remote-tracking branch 'doctrine/master' into DBAL-1168-add-default-schema-name

garex committed:
.. phpcs fix in test

deeky666 commented:
@garex sorry for the delay. Haven't been around here last year very much. The concern I have about the "default schema" topic how it is implemented in DBAL (via platforms) is that there really is no "fixed" default schema for probably all the vendors supporting schemas. In PostgreSQL for example you can configure the default schema via its search path, Oracle uses the user that connects as the default schema and I suppose others like SQL Server let the user define the default schema, too. All that happens at runtime so having the platforms return a default schema name is quite of a misconception IMO and should even be deprecated. Also the term "schema" and "namespace" as used in Doctrine is rather ambiguous and not clearly defined. IMO "schema" / "namespace" is a concept that groups database objects like tables, indexes etc. So in MySQL for example this concept would translate to "database".

So what we have to do (rather than relying on the platforms default schema name) is evaluate the runtime default schema name and add it to the table name when creating the schema form entity classes.

The problem IMO is [this line](https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Schema/Schema.php#L96). While assuming that `public` is the default namespace for all vendors (even those not supporting schemas) is completely wrong, the runtime default name [should already be evaluated in the schema manager](https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L1055-L1058).

So maybe this needs reevaluation. I don't think this solution is the best we can achieve.

garex commented:
Hi @deeky666 , nice to see you here.

I think it will be overcomplicated fix that will be merged may be in next century. Current solution is not best, but it will fix issue in 99% of cases. 

How often you see users, who change default schema names or default root name? I think it's a rare case.

Anyway if you want more ideal solution for 1% of users, it could be later improved by evaluating from live connection during schema diff. But as I assume it will be low-priority task that will be never done.

deeky666 commented:
@garex I get your point and you are possibly completely right about that it fixes the issue for "99%" of the users but I think we already have the possibility to fix this more properly without a lot of effort. The schema manager already has an API for evaluating the runtime schema. The reason I don't like the solution is that it pushes a misconception even further into the wrong direction. Also adding a special visitor for this makes us having to maintain that API which in fact IMO is just a workaround that doesn't even fix all edge cases. Speaking of edge cases, there is murphy's law and from my experience the "1%" edge cases will be requested to be fixed soon after that. So introducing this solution will only put us more into the corner.
I could probably live with a solution like this if in fact we do not have another better way of fixing this issue. I am only asking for evaluating if we can use the schema managers `getSchemaSearchPaths()` for this (because this is exactly the reasons why we implemented it IIRC). And looking at the code it should already do its job. IMO the logic should be:

1. evaluate the default schema via `getSchemaSearchPaths()`
2. if that does not return anything, use the default schema defined in the platform

That would save us an additional visitor, if I am not completely wrong.

garex commented:
@deeky666 I will research it and ping back here.

antonmedv commented:
@deeky666 any plans on solving this?

garex commented:
@antonmedv you can try to fix it yourself. @deeky666 is only as a reviewer here, not implementer/fixer.

greg0ire commented:
`getSchemaSearchPaths()` is no longer an option because of https://github.com/doctrine/dbal/pull/4821 `PostgreSQLSchemaManager::getCurrentSchema()` should be used instead.
